### PR TITLE
 [processing] Avoid field collision via optional prefix in overlay algorithms

### PR DESCRIPTION
**Pull request by nirvn**
## Description
Because appending _{n} to colliding field name when merging sourceA and sourceB's fields is not as predictable as declaring a prefix (which can be re-used in models, etc.).

## Checklist

> Reviewing is a process done by project maintainers, mostly on a volunteer basis. We try to keep the overhead as small as possible and appreciate if you help us to do so by completing the following items. Feel free to ask in a comment if you have troubles with any of them.

- [x] Commit messages are descriptive and explain the rationale for changes
- [x] Commits which fix bugs include `fixes #11111` in the commit message next to the description
- [x] Commits which add new features are tagged with `[FEATURE]` in the commit message
- [x] Commits which change the UI or existing user workflows are tagged with `[needs-docs]` in the commit message and contain sufficient information in the commit message to be documented
- [x] I have read the [QGIS Coding Standards](https://docs.qgis.org/testing/en/docs/developers_guide/codingstandards.html) and this PR complies with them
- [x] This PR passes all existing unit tests (test results will be reported by travis-ci after opening this PR)
- [x] New unit tests have been added for core changes
- [x] I have run [the `scripts/prepare-commit.sh` script](https://github.com/qgis/QGIS/blob/master/.github/CONTRIBUTING.md#contributing-to-qgis) before each commit

### CODE REVIEW

File: src/analysis/processing/qgsalgorithmintersection.cpp [hunk 1]
```diff
@@ -68,6 +68,8 @@ void QgsIntersectionAlgorithm::initAlgorithm( const QVariantMap & )
                   QObject::tr( "Overlay fields to keep (leave empty to keep all fields)" ), QVariant(),
                   QStringLiteral( "OVERLAY" ), QgsProcessingParameterField::Any, true, true ) );
 
+  addParameter( new QgsProcessingParameterString( QStringLiteral( "OVERLAY_FIELDS_PREFIX" ), QObject::tr( "Overlay fields prefix" ), QString( "" ), false, true ) );
```
Comment (nyalldawson) [inline src/analysis/processing/qgsalgorithmintersection.cpp]:
```suggestion
  addParameter( new QgsProcessingParameterString( QStringLiteral( "OVERLAY_FIELDS_PREFIX" ), QObject::tr( "Overlay fields prefix" ), QString(), false, true ) );
```

Comment (nyalldawson) [reply src/analysis/processing/qgsalgorithmintersection.cpp]: Should this be a flagged as an advanced parameter? (i think so)

File: src/analysis/processing/qgsalgorithmlineintersection.cpp [hunk 2]
```diff
@@ -63,6 +63,8 @@ void QgsLineIntersectionAlgorithm::initAlgorithm( const QVariantMap & )
                   QStringLiteral( "INTERSECT" ), QgsProcessingParameterField::Any,
                   true, true ) );
 
+  addParameter( new QgsProcessingParameterString( QStringLiteral( "INTERSECT_FIELDS_PREFIX" ), QObject::tr( "Intersect fields prefix" ), QString( "" ), false, true ) );
```
Comment (nyalldawson) [inline src/analysis/processing/qgsalgorithmlineintersection.cpp]:
```suggestion
  addParameter( new QgsProcessingParameterString( QStringLiteral( "INTERSECT_FIELDS_PREFIX" ), QObject::tr( "Intersect fields prefix" ), QString(), false, true ) );
```

File: src/analysis/processing/qgsalgorithmsymmetricaldifference.cpp [hunk 3]
```diff
@@ -55,6 +55,8 @@ void QgsSymmetricalDifferenceAlgorithm::initAlgorithm( const QVariantMap & )
 {
   addParameter( new QgsProcessingParameterFeatureSource( QStringLiteral( "INPUT" ), QObject::tr( "Input layer" ) ) );
   addParameter( new QgsProcessingParameterFeatureSource( QStringLiteral( "OVERLAY" ), QObject::tr( "Overlay layer" ) ) );
+  addParameter( new QgsProcessingParameterString( QStringLiteral( "OVERLAY_FIELDS_PREFIX" ), QObject::tr( "Overlay fields prefix" ), QString( "" ), false, true ) );
```
Comment (nyalldawson) [inline src/analysis/processing/qgsalgorithmsymmetricaldifference.cpp]:
```suggestion
  addParameter( new QgsProcessingParameterString( QStringLiteral( "OVERLAY_FIELDS_PREFIX" ), QObject::tr( "Overlay fields prefix" ), QString(), false, true ) );
```

File: src/analysis/processing/qgsalgorithmunion.cpp [hunk 4]
```diff
@@ -61,6 +61,7 @@ void QgsUnionAlgorithm::initAlgorithm( const QVariantMap & )
 {
   addParameter( new QgsProcessingParameterFeatureSource( QStringLiteral( "INPUT" ), QObject::tr( "Input layer" ) ) );
   addParameter( new QgsProcessingParameterFeatureSource( QStringLiteral( "OVERLAY" ), QObject::tr( "Overlay layer" ), QList< int >(), QVariant(), true ) );
+  addParameter( new QgsProcessingParameterString( QStringLiteral( "OVERLAY_FIELDS_PREFIX" ), QObject::tr( "Overlay fields prefix" ), QString( "" ), false, true ) );
```
Comment (nyalldawson) [inline src/analysis/processing/qgsalgorithmunion.cpp]:
```suggestion
  addParameter( new QgsProcessingParameterString( QStringLiteral( "OVERLAY_FIELDS_PREFIX" ), QObject::tr( "Overlay fields prefix" ), QString(), false, true ) );
```

**Review by nyalldawson — COMMENTED** [review]
Looks good, just needs tests

**Comment by DelazJ** [comment]
@nirvn should this [needs-docs]?

File: src/core/processing/qgsprocessingutils.h [hunk 5]
```diff
@@ -287,7 +287,7 @@ class CORE_EXPORT QgsProcessingUtils
      * length of field names, so be aware that the results of calling this method may
      * be truncated when saving to these formats.
      */
-    static QgsFields combineFields( const QgsFields &fieldsA, const QgsFields &fieldsB );
+    static QgsFields combineFields( const QgsFields &fieldsA, const QgsFields &fieldsB, const QString fieldsBPrefix = QString() );
```
Comment (m-kuhn) [inline src/core/processing/qgsprocessingutils.h]:
```suggestion
    static QgsFields combineFields( const QgsFields &fieldsA, const QgsFields &fieldsB, const QString &fieldsBPrefix = QString() );
```

**Pull request merged**